### DISCUSSION TITLE
STM32411 (Blackpill) and RS485

### DISCUSSION THREAD

**Discussion by rvalotta** [terjeio/grblHAL#330] [Q&A] [answered]
Just got a VFD and wanted to know if i could use a RS385 module (http://www.yourduino.com/sunshop/index.php?l=product_detail&p=323#:~:text=This%20module%20interfaces%20an%20Arduino,the%20same%20RS%2D485%20wiring.) to convert the TTL to rs485? Where would I connect the RE and DE? I know tx and rx are DI and RI.

**Comment by terjeio** [comment]
You can use it provided that support for the direction signal is added to the driver. The easiest way would be to let the [modbus plugin ](https://github.com/grblHAL/Plugins_spindle) claim the highest numbered auxillary output (if available) on init, a bit more involved would be to add it to the _driver.c_ code.
Easier still is to get a module that has auto direction support...

> Where would I connect the RE and DE?

Short them together and connect to the direction signal pin.

Note that this repo is no longer updated, the active one is [here](https://github.com/grblHAL).

**Reply by rvalotta** [reply]
Thanks I'm aware the repo is at the other location but the discussion board is over here (at least I don't see a link at the top of the current repo)

**Comment by dresco** [comment]
@rvalotta  

Fwiw, I'm using ones like these - https://www.aliexpress.com/item/32896018384.html
Dirt cheap from multiple sellers & work fine from a 3.3v supply.

**Comment by rvalotta** [comment]
Okay so I've setup the VFD and gotten it working. Now I'm working on compiling VFD support into GRBL Hal and have some questions

When defining SPINDLE_HUANYANG, What is the difference between 1 and 2
Do I also need to include MODBUS? 

Can I comment out and reclaim by Spindle En/dir/pwm pins?

For shits and giggle I tried compiling it with SPINDLE_HUANYANG 2 and  MODBUS_ENABLE        1 and am getting the following errors:

> Src\driver.c: In function 'selectStream':
Src\driver.c:384:19: error: 'io_stream_t {aka struct <anonymous>}' has no member named 'disable_rx'; did you mean 'disable'?
     if(hal.stream.disable_rx)
                   ^~~~~~~~~~
                   disable
Src\driver.c:385:20: error: 'io_stream_t {aka struct <anonymous>}' has no member named 'disable_rx'; did you mean 'disable'?
         hal.stream.disable_rx(false);
                    ^~~~~~~~~~
                    disable
Src\driver.c: In function 'driver_init':
Src\driver.c:1956:5: error: too few arguments to function 'huanyang_init'
     huanyang_init();
     ^~~~~~~~~~~~~
In file included from Inc/driver.h:267:0,
                 from Src\driver.c:28:
./spindle/huanyang.h:42:6: note: declared here
 void huanyang_init (modbus_stream_t *stream);
      ^~~~~~~~~~~~~
*** [.pio\build\black_pill_411\src\driver.o] Error 1
Src\serial.c: In function 'serialInit':
Src\serial.c:217:10: error: 'io_stream_t {aka const struct <anonymous>}' has no member named 'disable_rx'; did you mean 'disable'?
         .disable_rx = serialDisable,
          ^~~~~~~~~~
          disable
Src\serial.c:217:23: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
         .disable_rx = serialDisable,
                       ^~~~~~~~~~~~~
Src\serial.c:217:23: note: (near initialization for 'stream.write_n')
Src\serial.c: In function 'serial2Init':
Src\serial.c:495:10: error: 'io_stream_t {aka const struct <anonymous>}' has no member named 'disable_rx'; did you mean 'disable'?
         .disable_rx = serial2Disable,
          ^~~~~~~~~~
          disable
Src\serial.c:495:23: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
         .disable_rx = serial2Disable,
                       ^~~~~~~~~~~~~~
Src\serial.c:495:23: note: (near initialization for 'stream.write_n')
*** [.pio\build\black_pill_411\src\serial.o] Error 1

Guessing that there are a few things that need updating. Is there a working driver that I can look at to reference what needs changing?

Thanks in advance.

**Comment by terjeio** [comment, marked as answer]
> When defining SPINDLE_HUANYANG, What is the difference between 1 and 2

The protocol version, 1 is not following Modbus conventions, 2 is. I do not know model numbers vs. protocol version, can be found in the VFD driver documentation?

> Do I also need to include MODBUS?

No, it is done automatically.

> Can I comment out and reclaim by Spindle En/dir/pwm pins?

Yes.

> For shits and giggle I tried compiling it with SPINDLE_HUANYANG 2 and MODBUS_ENABLE 1 and am getting the following errors:

You have to [download the submodules](https://github.com/grblHAL/core/wiki/Compiling-GrblHAL#using-git-command-line) along with the driver code when updating.

> Is there a working driver that I can look at to reference what needs changing?

The current driver should work when the submodules are in sync with the driver.

**Reply by rvalotta** [reply]
DOH forgot the --recurse-submodules... Your right its compiling now.

How do i find out what pins its using for modbus. I think its Serial.h where its setting the RX and TX pins but i'm not sure

**Reply by terjeio** [reply]
Currently via checking the source. Not good I know, I plan to add non GPIO pins to the `$pins` report.
For the 411 Blackpill it is PA2 and PA3.

**Reply by rvalotta** [reply]
Thankyou I appreciate that. Could you tell me which file the above is defined in. Its more for me to learn where things are...

**Reply by terjeio** [reply]
In [serial.c](https://github.com/grblHAL/STM32F4xx/blob/master/Src/serial.c) - pins are assigned based on `IS_NUCLEO_DEVKIT` and `SERIAL2_MOD` symbols defined elsewhere. Search with CTRL+H to find them.

**Reply by dresco** [reply]
> > When defining SPINDLE_HUANYANG, What is the difference between 1 and 2
> 
> The protocol version, 1 is not following Modbus conventions, 2 is. I do not know model numbers vs. protocol version, can be found in the VFD driver documentation?

As far as I know, all the HY series are v1 protocol, and the GT series (newer, with vector control) are v2 protocol..
